File: packages/intellij-plugin/engine/src/test/kotlin/dev/bredson/tokentracker/engine/EngineTest.kt
package dev.bredson.tokentracker.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

private class FakeSource(
    private val available: Boolean,
    private val initial: List<UsageRecord>,
    var titles: Map<String, String> = emptyMap(),
    var unknown: List<String> = emptyList(),
) : UsageSource {
    override val id = "fake"
    override val displayName = "Fake source"
    private var callback: ((List<UsageRecord>) -> Unit)? = null
    var disposed = false
    var loadCount = 0

    override fun detect() = available
    override fun loadAll(): List<UsageRecord> { loadCount++; return initial }
    override fun watch(onUpdate: (List<UsageRecord>) -> Unit): Disposable {
        callback = onUpdate
        return Disposable { disposed = true }
    }
    fun emit(records: List<UsageRecord>) = callback?.invoke(records)
    override fun sessionTitles() = titles
    override fun unknownModels() = unknown
}

class EngineTest {
    @Test
    fun `start loads initial records only from sources that detect as available`() {
        val engine = Engine(listOf(FakeSource(true, listOf(record(id = "r1"))), FakeSource(false, listOf(record(id = "r2")))))
        engine.start()
        assertEquals(listOf("r1"), engine.records().map { it.id })
    }

    @Test
    fun `start is idempotent`() {
        val source = FakeSource(true, listOf(record(id = "r1")))
        val engine = Engine(listOf(source))
        engine.start(); engine.start()
        assertEquals(1, engine.records().size)
        assertEquals(1, source.loadCount)
    }

    @Test
    fun `records emitted by watch are appended`() {
        val source = FakeSource(true, listOf(record(id = "r1")))
        val engine = Engine(listOf(source)).apply { start() }
        source.emit(listOf(record(id = "r2")))
        assertEquals(listOf("r1", "r2"), engine.records().map { it.id }.sorted())
    }

    @Test
    fun `records are deduplicated by id`() {
        val source = FakeSource(true, listOf(record(id = "r1")))
        val engine = Engine(listOf(source)).apply { start() }
        source.emit(listOf(record(id = "r1")))
        assertEquals(1, engine.records().size)
    }

    @Test
    fun `listeners are notified only when something new arrived`() {
        val source = FakeSource(true, listOf(record(id = "r1")))
        val engine = Engine(listOf(source)).apply { start() }
        var notifications = 0
        engine.onChange { notifications++ }
        source.emit(listOf(record(id = "r1")))
        assertEquals(0, notifications)
        source.emit(listOf(record(id = "r2")))
        assertEquals(1, notifications)
    }

    @Test
    fun `an explicitly empty batch is a metadata change and notifies listeners`() {
        val source = FakeSource(true, emptyList())
        val engine = Engine(listOf(source)).apply { start() }
        var notifications = 0
        engine.onChange { notifications++ }
        source.emit(emptyList())
        assertEquals(1, notifications)
        assertTrue(engine.records().isEmpty())
    }

    @Test
    fun `a disposed subscription is no longer notified`() {
        val source = FakeSource(true, emptyList())
        val engine = Engine(listOf(source)).apply { start() }
        var notifications = 0
        engine.onChange { notifications++ }.dispose()
        source.emit(listOf(record(id = "r1")))
        assertEquals(0, notifications)
    }

    @Test
    fun `aggregations reflect the current store`() {
        val engine = Engine(
            listOf(
                FakeSource(
                    true,
                    listOf(
                        record(id = "r1", taskId = "t1", sessionId = "s1", projectPath = "/p1", costUsd = 1.0),
                        record(id = "r2", taskId = "t1", sessionId = "s1", projectPath = "/p1", costUsd = 2.0),
                    ),
                ),
            ),
        ).apply { start() }
        assertEquals(3.0, engine.aggregateByTask().single().totals.costUsd, 1e-9)
        assertEquals("s1", engine.aggregateBySession().single().key)
        assertEquals("/p1", engine.aggregateByProject().single().key)
        assertEquals("2026-03-10", engine.aggregateByPeriod(PeriodGranularity.DAY).single().key)
    }

    @Test
    fun `session titles are merged across sources, later source wins`() {
        val first = FakeSource(true, emptyList(), titles = mapOf("s1" to "Pierwszy"))
        val second = FakeSource(true, emptyList(), titles = mapOf("s1" to "Drugi", "s2" to "Inny"))
        val engine = Engine(listOf(first, second)).apply { start() }
        assertEquals(mapOf("s1" to "Drugi", "s2" to "Inny"), engine.sessionTitles())
    }

    @Test
    fun `unknown models are merged, deduplicated and sorted`() {
        val first = FakeSource(true, emptyList(), unknown = listOf("zeta", "alpha"))
        val second = FakeSource(true, emptyList(), unknown = listOf("alpha", "mid"))
        val engine = Engine(listOf(first, second)).apply { start() }
        assertEquals(listOf("alpha", "mid", "zeta"), engine.unknownModels())
    }

    @Test
    fun `dispose stops watches and listeners`() {
        val source = FakeSource(true, emptyList())
        val engine = Engine(listOf(source)).apply { start() }
        var notifications = 0
        engine.onChange { notifications++ }
        engine.dispose()
        assertTrue(source.disposed)
        source.emit(listOf(record(id = "r1")))
        assertEquals(0, notifications)
    }
}

File: packages/engine/test/engine.test.ts
import { describe, expect, it, vi } from "vitest";
import { Engine } from "../src/engine";
import type { Disposable, UsageRecord, UsageSource } from "../src/types";

function makeRecord(overrides: Partial<UsageRecord> = {}): UsageRecord {
  return {
    id: "rec-1",
    source: "fake",
    sessionId: "session-1",
    taskId: "task-1",
    timestamp: "2026-01-01T00:00:00.000Z",
    model: "claude-sonnet-5",
    tokensInput: 100,
    tokensOutput: 50,
    tokensCacheRead: 0,
    tokensCacheWrite: 0,
    costUsd: 1,
    projectPath: "/tmp/project",
    ...overrides,
  };
}

class FakeSource implements UsageSource {
  readonly id = "fake";
  readonly displayName = "Fake source";

  private watchCallback: ((records: UsageRecord[]) => void) | undefined;
  disposed = false;

  constructor(
    private readonly available: boolean,
    private readonly initialRecords: UsageRecord[],
  ) {}

  async detect(): Promise<boolean> {
    return this.available;
  }

  async loadAll(): Promise<UsageRecord[]> {
    return this.initialRecords;
  }

  watch(onUpdate: (newRecords: UsageRecord[]) => void): Disposable {
    this.watchCallback = onUpdate;
    return {
      dispose: () => {
        this.disposed = true;
      },
    };
  }

  /** Test helper: simulate the source's watcher pushing new records. */
  emit(records: UsageRecord[]): void {
    this.watchCallback?.(records);
  }

  getSessionTitles(): Map<string, string> {
    return this.titles ?? new Map();
  }

  titles: Map<string, string> | undefined;

  getUnknownModels(): string[] {
    return this.unknownModels ?? [];
  }

  unknownModels: string[] | undefined;
}

class MinimalSource implements UsageSource {
  readonly id = "minimal";
  readonly displayName = "Minimal source";
  async detect(): Promise<boolean> {
    return true;
  }
  async loadAll(): Promise<UsageRecord[]> {
    return [];
  }
  watch(): Disposable {
    return { dispose: () => {} };
  }
}

describe("Engine.start", () => {
  it("loads initial records only from sources that detect() as available", async () => {
    const available = new FakeSource(true, [makeRecord({ id: "r1" })]);
    const unavailable = new FakeSource(false, [makeRecord({ id: "r2" })]);

    const engine = new Engine({ sources: [available, unavailable] });
    await engine.start();

    expect(engine.getRecords().map((r) => r.id)).toEqual(["r1"]);
  });

  it("is idempotent — calling start() twice does not reload or double-watch", async () => {
    const source = new FakeSource(true, [makeRecord({ id: "r1" })]);
    const engine = new Engine({ sources: [source] });

    await engine.start();
    await engine.start();

    expect(engine.getRecords()).toHaveLength(1);
  });
});

describe("Engine ingest via watch()", () => {
  it("appends newly emitted records to the store", async () => {
    const source = new FakeSource(true, [makeRecord({ id: "r1" })]);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    source.emit([makeRecord({ id: "r2" })]);

    expect(engine.getRecords().map((r) => r.id).sort()).toEqual(["r1", "r2"]);
  });

  it("deduplicates records by id, ignoring re-emitted duplicates", async () => {
    const source = new FakeSource(true, [makeRecord({ id: "r1" })]);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    source.emit([makeRecord({ id: "r1" })]);

    expect(engine.getRecords()).toHaveLength(1);
  });

  it("notifies onChange listeners only when new records are actually added", async () => {
    const source = new FakeSource(true, [makeRecord({ id: "r1" })]);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    const listener = vi.fn();
    engine.onChange(listener);

    source.emit([makeRecord({ id: "r1" })]); // duplicate, should not notify
    expect(listener).not.toHaveBeenCalled();

    source.emit([makeRecord({ id: "r2" })]); // new, should notify
    expect(listener).toHaveBeenCalledTimes(1);
  });

  it("treats an explicitly empty batch as a metadata change and notifies listeners", async () => {
    const source = new FakeSource(true, []);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    let notifications = 0;
    engine.onChange(() => notifications++);

    source.emit([]);

    expect(notifications).toBe(1);
    expect(engine.getRecords()).toEqual([]);
  });

  it("stops notifying a listener after its subscription is disposed", async () => {
    const source = new FakeSource(true, []);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    const listener = vi.fn();
    const subscription = engine.onChange(listener);
    subscription.dispose();

    source.emit([makeRecord({ id: "r1" })]);
    expect(listener).not.toHaveBeenCalled();
  });
});

describe("Engine aggregation delegation", () => {
  it("aggregateByTask/Session/Project/Period reflect the current store", async () => {
    const source = new FakeSource(true, [
      makeRecord({ id: "r1", taskId: "t1", sessionId: "s1", projectPath: "/p1", costUsd: 1 }),
      makeRecord({ id: "r2", taskId: "t1", sessionId: "s1", projectPath: "/p1", costUsd: 2 }),
    ]);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    const byTask = engine.aggregateByTask();
    expect(byTask).toHaveLength(1);
    expect(byTask[0].totals.costUsd).toBe(3);

    const bySession = engine.aggregateBySession();
    expect(bySession).toHaveLength(1);
    expect(bySession[0].totals.recordCount).toBe(2);

    const byProject = engine.aggregateByProject();
    expect(byProject).toHaveLength(1);

    const byDay = engine.aggregateByPeriod("day");
    expect(byDay).toHaveLength(1);
  });
});

describe("Engine.getSessionTitles", () => {
  it("returns an empty map when no source implements getSessionTitles", async () => {
    const source = new FakeSource(true, []);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    expect(engine.getSessionTitles().size).toBe(0);
  });

  it("merges titles from sources that implement getSessionTitles", async () => {
    const source = new FakeSource(true, []);
    source.titles = new Map([["session-1", "Tytuł sesji"]]);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    expect(engine.getSessionTitles().get("session-1")).toBe("Tytuł sesji");
  });

  it("lets a later source override an earlier source's title for the same session", async () => {
    const first = new FakeSource(true, []);
    first.titles = new Map([["session-1", "Pierwszy tytuł"]]);
    const second = new FakeSource(true, []);
    second.titles = new Map([["session-1", "Drugi tytuł"]]);
    const engine = new Engine({ sources: [first, second] });
    await engine.start();

    expect(engine.getSessionTitles().get("session-1")).toBe("Drugi tytuł");
  });
});

describe("Engine.getUnknownModels", () => {
  it("returns an empty list when no source implements getUnknownModels", async () => {
    const engine = new Engine({ sources: [new MinimalSource()] });
    await engine.start();
    expect(engine.getUnknownModels()).toEqual([]);
  });

  it("merges and deduplicates unknown models across sources, sorted", async () => {
    const first = new FakeSource(true, []);
    first.unknownModels = ["zeta", "alpha"];
    const second = new FakeSource(true, []);
    second.unknownModels = ["alpha", "mid"];
    const engine = new Engine({ sources: [first, second] });
    await engine.start();
    expect(engine.getUnknownModels()).toEqual(["alpha", "mid", "zeta"]);
  });
});

describe("Engine.dispose", () => {
  it("disposes every source's watch subscription", async () => {
    const source = new FakeSource(true, []);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    engine.dispose();

    expect(source.disposed).toBe(true);
  });

  it("stops notifying listeners after dispose", async () => {
    const source = new FakeSource(true, []);
    const engine = new Engine({ sources: [source] });
    await engine.start();

    const listener = vi.fn();
    engine.onChange(listener);
    engine.dispose();

    source.emit([makeRecord({ id: "r1" })]);
    expect(listener).not.toHaveBeenCalled();
  });
});
